### ISSUE TITLE
Make IDistributedCache an IDisposable

### ISSUE THREAD

**Issue by njy** [dotnet/runtime#44108]
### Description

A MemoryDistributedCache internally creates (in the ctor) an instance of an IMemoryCache as a backing store, and that makes sense.

https://github.com/dotnet/runtime/blob/802543648478f9b0b9f7e07ee9bbde823cd90a64/src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryDistributedCache.cs#L33

But since IMemoryCache implements IDisposable and it should be disposed when no more needed, and since neither the IDistributedCache interface nor the MemoryDistributedCache concrete type implement IDisposable, this is simply impossible.

### Configuration

Any

### Proposal
I propose to have either IDistributedCache or MemoryDistributedCache implement IDisposable so that it is possible not to leak memory, and for what it's worth I would definitely go with the change directly to IDistributedCache interface and since it would be a breaking change we are probably talking about the next major release train, like .NET 6.

I can imagine having such a change (more so in the IDistributedCache interface, less so in the MemoryDistributedCache implementation) may be a delicate thing, but I really can't see any other way around this.

If one of the maintainers thinks that this would be ok, I would be glad to create a PR for that.

Please let me know before creating a useless PR that you would then have to follow along.

**Comment by KalleOlaviNiemitalo** [comment]
That is not a leak. MemoryCache.Dispose would not release any resources, anyway:

<https://github.com/dotnet/runtime/blob/802543648478f9b0b9f7e07ee9bbde823cd90a64/src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs#L484-L500>

MemoryDistributedCache is in the same assembly as MemoryCache, so MemoryDistributedCache can assume that MemoryCache behaves this way. A class derived from MemoryCache might behave differently but MemoryDistributedCache provides no way to inject such a class.

**Comment by njy** [comment]
At a second look yes, you are absolutely right: in the specific case of MemoryDistributedCache that is the case.

More generaly though I'm wondering, if a "local" interface like IMemoryCache is also IDisposable, why that is not the case with a "not local" one like IDistributedCache that, very much likely, will need to dispose connections or something like that as we can see in the [RedisCache ](https://github.com/dotnet/extensions/blob/master/src/Caching/StackExchangeRedis/src/RedisCache.cs).

Now, of course the RedisCache class can easily implement IDisposable nonetheless (and in fact it does) but I'm thinking about a scenario like this:

``` csharp
IDistributedCache CreateDistributedCache() {
  ...
}

var cache = CreateDistributedCache();
```

and how I would be able to handle the disposal correctly, for example in the case of a RedisCache.

Basically, IDistributedCache not being an IDisposable means not being able to dispose in all possible scenarios, and that would be a problem.
Instead, by having IDistributedCache being an IDisposable, at most you would have a specific distributed cache impl not needing it having just a (very small) hassle, which would be easily done in 2 lines like in MemoryCache impl as you have correctly pointed out above.

Can you please help me understand the rationale behind such decision or if you think this change would make sense?

Thanks.

**Comment by njy** [comment]
(I've changed the issue title to better reflect the discussion)

**Comment by KalleOlaviNiemitalo** [comment]
I guess the cache was intended to be disposed by a dependency injection container, not by types that use the IDistributedCache service. The AddDistributedMemoryCache method adds MemoryDistributedCache as a singleton, so if many types in an application need IDistributedCache, the DI container gives the same MemoryDistributedCache instance to each of them. Then, those types should not call Dispose because they don't know who else might still be using the same instance. Keeping the Dispose method a bit hidden may help developers comply with this.

<https://github.com/dotnet/runtime/blob/802543648478f9b0b9f7e07ee9bbde823cd90a64/src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCacheServiceCollectionExtensions.cs#L84>

If you don't use dependency injection, you can do things like these in C# anyway:

```C#
var cache = CreateDistributedCache();
try
{
    // use the cache
}
finally
{
    (cache as IDisposable)?.Dispose();
}
```

or:

```C#
var cache = CreateDistributedCache();
using (cache as IDisposable)
{
    // use the cache
}
```

so, I don't think making IDistributedCache require IDisposable would make appropriate disposal so much easier that it would justify the breaking change.

**Comment by KalleOlaviNiemitalo** [comment]
> if a "local" interface like IMemoryCache is also IDisposable, why that is not the case with a "not local" one like IDistributedCache

Perhaps that has something to do with IDistributedCache only supporting `string` keys and `byte[]` values, and no eviction callbacks.

**Comment by njy** [comment]
> I guess the cache was intended to be disposed by a dependency injection container, not by types that use the IDistributedCache service.

Got it, I'm seeing this more and more: the dependency on dependency injection (no pun intended) in .net core as if that is **the** way to do it and not **a** way to do it (see http client & friends).
I'm not that happy about it, but I think I get the rationale behind it.

> The AddDistributedMemoryCache method adds MemoryDistributedCache as a singleton, so if many types in an application need IDistributedCache, the DI container gives the same MemoryDistributedCache instance to each of them. Then, those types should not call Dispose because they don't know who else might still be using the same instance. Keeping the Dispose method a bit hidden may help developers comply with this.

Yes, but:
1) this is because the MemoryDistributedCache impl is to be used as a singleton, whereas other impl may need to work differently, so we are binding the overall usage to a specific impl, I think. If I would like to create an impl that is scoped or transient I would not be able to do that (withtout hacks) most probably, right?
2) the same reasoning should be valid for IMemoryCache, but it is not since it explicitly is an IDisposable

I'd like to be clear here: I'm not insisting to troll or something, I hope that is clear.
I'm really just trying to understand the rationale behind the approach (and maybe see if a change may make sense) and how to apply these concepts to be a better .net citizen, for example when desiging and releasing a lib for other peeople to use on this - stupendous, let me say - platform.
Even if the change would not go along because of backward compat, at least I'd like to understand if the change itself would make sense.

> If you don't use dependency injection, you can do things like these in C# anyway:
> 
> ```cs
> var cache = CreateDistributedCache();
> try
> {
>     // use the cache
> }
> finally
> {
>     (cache as IDisposable)?.Dispose();
> }
> ```

Yes, in fact that is what I'm currently using but it's hacky and cannot use the standard `using` pattern.

> Perhaps that has something to do with IDistributedCache only supporting string keys and byte[] values, and no eviction callbacks.

I thought about that, but as you pointed out above the Dispose method in MemoryCache [does nothing basically](https://github.com/dotnet/runtime/blob/master/src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs#L484-L500) so IDisposable seems not to be needed in fact.
It does not remove all the entries, it does not call post eviction callbacks on them (if any) and so on, which is another thing that surprised me: at first I thought that disposing a memory cache would mean all entries being removed and post eviction callbacks beening called, but looking at the source that is not the case at all.
Imho disposing the entire cache should be _logically_ similar to removing all of them one by one.

Also, following the same reasoning, if IMemoryCache is IDisposable because of eviction callbacks, IDistributedCache should be even more so because of remote connections handling or similar.

> so, I don't think making IDistributedCache require IDisposable would make appropriate disposal so much easier that it would justify the breaking change.

It seems reasonable to me for the reasons above, but I can see the fear of a breaking change.
Keeping aside the breaking nature of the change for a moment, if the proposed change would seem reasonable then maybe there's the possibility to involve the majority of IDistributedCache implementers (not that many if I'm right) and see what they prefer, and I suspect most of them are already implementing IDisposable anyway.

Would this make sense?


ps: thanks again for taking the time to discuss and explain, really appreciate that.

**Comment by davidfowl** [comment]
The implementation should be disposable not the interface

**Comment by njy** [comment]
@davidfowl why not the interface though, like IMemoryCache? Is there a particular reason?
Thanks.

**Comment by eerhardt** [comment]
@Tratcher - is there a reasoning behind this difference between `IMemoryCache` and `IDistributedCache`? I see it has been this way since these 2 interfaces were introduced six years ago.

https://github.com/aspnet/Caching/pull/2
https://github.com/aspnet/Caching/pull/23

**Comment by Tratcher** [comment]
I don't remember, I don't even see any reasons that the current implementation needs to be IDisposable. It may be related to how we attempted automatic compaction in earlier implementations.

But @davidfowl's right, from a DI perspective only the implementation needs to implement IDisposable, not the interface.

**Comment by njy** [comment]
> from @davidfowl 
> The implementation should be disposable not the interface
> [...]
> from @Tratcher 
> I don't remember, I don't even see any reasons that the current implementation needs to be IDisposable. It may be related to how we attempted automatic compaction in earlier implementations.
> But @davidfowl's right, from a DI perspective only the implementation needs to implement IDisposable, not the interface.

I absolutely agree _in theory_, but since I was trying to see how to play well with the rest of the framework and IMemoryCache (the interface) is also directly IDisposable I thought that there was a reason, and the same reason would then apply to IDistributedCache, too.

Since it seems clear now that there is no actual reason and it was probably just an old attempt to better control auto compaction or something and it is still there probably for backward compatibilty (I would assume) then I agree it does not make sense to have it in IDistributedCache, too.

One last thing out of pure curiosity: even though it may be complicated to handle such a case while executing the `Dispose` method, would it be right to assume that - at least in theory - disposing a memory means logically removing all the entries? And if so, shouldn't it, again at least in theory, also call all the remaining entries' post eviction callbacks (if any)?
I mean, the expectations of someone adding an entry to a cache with a post eviction callback is to be able to do something when the entriy will get out of the cache, so disposing the entire cache would logically match such a case.

**Just to be clear**: I'm not proposing to do it, just trying to understand it the theory behind such a thing would make sense.

Apart from this I think we can close this issue.

And thanks for the explanations.

**Comment by eerhardt** [comment]
> disposing a memory means logically removing all the entries? And if so, shouldn't it, again at least in theory, also call all the remaining entries' post eviction callbacks (if any)?

I guess it depends on what you expect `Dispose()` does. To me, calling `Dispose()` means I'm finished using this object and it should release any unmanaged resources. I guess I wouldn't expect it to logically `Clear()` a collection. So I wouldn't expect more events to fire from a cache that was being disposed.

**Comment by njy** [comment]
> To me, calling Dispose() means I'm finished using this object and it should release any unmanaged resources.

Ah, that's interesting. To me is kinda the same, but also it should release any _managed_ resources too, including the ones that they have "transitively".
Now some of them may be inside cache entries and to be disposed their "owner" (should be notified via a post eviction callback, just like it would be done by removing the entries manually.
I suppose the expectation (or better, probably **my** expectation) is that when you put something in a memory cache with a post eviction callback is like saying "hey, put this value for key "foo" and call be back when it will be evicted for whatever reason" of course with the reason being specified with the [EvictionReason enum](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.caching.memory.evictionreason?view=dotnet-plat-ext-3.1).

Sometimes the value in the entries are disposables that need disposing when evicted otherwise memory leaks or similar, and not being able to know when an entire cache is being disposed + not being notified when all of the entries will be "discarded" makes this situation (which btw I'm experiencing right now in a project) basically impossible to handle correctly.

> I guess I wouldn't expect it to logically Clear() a collection. So I wouldn't expect more events to fire from a cache that was being disposed.

If you think about a cache just like a concurrent dictionary or as a collection in general then yes, I can see why and I would agree, too.

But add post eviction callbacks to the mix and I would say the opposite would be a reasonable expectation, otherwise instead of saying _"take this and call me back when it will be evicted since I have cleanup to do"_ it will be like saying _"take this and call me back when it will be evicted since I have cleanup to do, unless you are evicting all of them then I don't care"_ which I would honestly don't understand.

And yes, usually an IMemoryCache is used with a MemoryCache impl with a singleton lifetime so disposing it means the application is most probably shutting down anyway, but that is just one of the possible implementations and lifetimes combo, it's not the only one.

Anyway, just my 2 cents.

**Comment by njy** [comment]
Also: in my current scenario I have complete control over the memory cache instance, and I control when it will be disposed, and I can do pretty much whatever I want.

Nonetheless I cannot correctly dispose the remaining entries in the cache before disposing the cache itself because the memory cache does not provide a way to iterate on all entries, so I cannot do proper cleanup.

I thought about creepy hack: keeping a separate concurrent bag with all the key that have been added to the memory cache and - just before disposing the cache - iterate over the concurrent bag to get all the key potentially still in the cache (if not already eviceted) and try to get the related entries from the cache and, if there, dispose them.

But, honestly: no, thanks but no.

Any idea?

**Comment by eerhardt** [comment]
This isn't an official suggestion, but if you really want the post eviction callback to fire, you could try calling `Compact(1.0)` on the cache before calling `Dispose()` (or override `Dispose(true)` and call `Compact`). This should cause the cache to be cleared and the callbacks to fire.

> the memory cache does not provide a way to iterate on all entries

See #36554 and #36026 for proposals in this area.

**Comment by njy** [comment]
Thanks @eerhardt I thought about that (I should've mentioned) but it will not work for entries with priority set to NeverRemove as you can see here:

https://github.com/dotnet/runtime/blob/master/src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs#L432

NeverRemove is intended (I assume) to avoid automatic eviction based on size, but in the case of "clearing the entire cache" it doesn't feit.

Anyway that could still be a good idea if one can control that no entry is used with priority set to NeverRemove.

**Comment by njy** [comment]
I'm closing this issue since in fact there's no reason to make the IDistributedCache interface an IDisposable directly.

I still believe that the Dispose method of MemoryCache should clean up all entries, mainly because of post eviction callbacks, but that would be for another issue.

Thanks everybody for participating in the discussion.
